homework/index.js

```diff
@@ -16,31 +42,86 @@
+      text: repos.name,
+    });
+
+    addRow(tbody, 'Description', repos.description);
+    addRow(tbody, 'Forks', repos.forks);
+    addRow(tbody, 'Updated', new Date(repos.updated_at).toLocaleString());
+  }
+
+  function renderContributors(contributorRepo, ulContributorInfo) {
+    const url = contributorRepo.contributors_url;
+    fetchJSON(url, (err, contributors) => {
+      if (err) {
+        renderError(err);
+        return;
       }
+      ulContributorInfo.innerHTML = '';
+
+      contributors.forEach(contributor => {
+        const contributorLink = createAndAppend('a', ulContributorInfo, {
+          href: contributor.html_url,
+          target: '_blank',
+        });
+        const listItem = createAndAppend('li', contributorLink, {
+          class: 'contributor-list-item',
+        });
+        createAndAppend('img', listItem, {
+          src: contributor.avatar_url,
+          alt: `${contributor.login} photo`,
+        });
+        createAndAppend('h1', listItem, { text: contributor.login });
+        createAndAppend('p', listItem, { text: contributor.contributions });
+      });
+    });
+  }
+
+  function createOptionElements(repositories, select) {
+    repositories.forEach((repo, index) => {
+      createAndAppend('option', select, { text: repo.name, value: index });
     });
-    return elem;
   }
 
   function main(url) {
+    const root = document.getElementById('root');
+    const header = createAndAppend('header', root);
+    const infoSection = createAndAppend('section', root, { class: 'wrapper' });
+    const headerWrapper = createAndAppend('div', header, { class: 'header-wrapper' });
+    createAndAppend('h1', headerWrapper, { text: 'HYF Repositories' });
+    const select = createAndAppend('select', headerWrapper);
+    const repoInfo = createAndAppend('div', infoSection, { class: 'flex-container' });
+    const contributorInfo = createAndAppend('ul', infoSection, {
+      class: 'flex-container',
+    });
+
     fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
       if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+        renderError(err);
         return;
       }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+      repositories.sort((a, b) => a.name.localeCompare(b.name));
+      createOptionElements(repositories, select);
+      const repo = repositories[select.value];
+      renderRepositories(repo, repoInfo);
+      renderContributors(repo, contributorInfo);
+
+      select.addEventListener('change', () => {
+        // const repo = repositories[select.value];
+        renderRepositories(repo, repoInfo);
+        renderContributors(repo, contributorInfo);
+      });
     });
   }
-
   const HYF_REPOS_URL = 'https://api.github.com/orgs/HackYourFuture/repos?per_page=100';
-  window.onload = () => main(HYF_REPOS_URL);
+  window.onload = main(HYF_REPOS_URL);
```

Comment:
I overlooked this change that you made which actually introduces a subtle bug. Now, the main function is directly called in 126 and its return value (`undefined`) is assigned to the `window.onload` property. You will not have noticed any issues, but if somebody decides to move script tag that loads this JavaScript file from the `body` to the `head` tag of the `index.htm` file your application will no longer run.

Try it, and see if you can explain what is happening.

Reply:
Luckily or unluckily I have made this mistake. So I watched and read about window.onload event after your review.  You are right. You made me thinking about that problem.  While I was writing window.onload it was automatic for me .But now I hope I understood what it means. So:
**Scenario 1: js inside the head tag. in Js file  ` window.onload = main(HYF_REPOS_URL)`;**
I got an error when I try to use js in head tag. Because Js reads files from top to bottom. So when it come  to script tag Js will not understand the code. Because `window.onload` property is actually null.
**Scenario 2: js in different file. but declared in body tag. `window.onload = main(HYF_REPOS_URL)`;**
This was my code. I did not get an error. Because  when js was reading body tag since my script tags were at the end of the body tag ,  Js understood what the main function means.
 Am I right?

Reply:
No, it is not quite like that. As soon as the browser encounters your script tag it will load the JavaScript file and will immediately start executing all executable code. Function _definitions_ are by themselves not directly executable, but function _calls_ are.

This is the erroneous code:

```js
window.onload = main(HYF_REPOS_URL);
```

You immediately call function `main` and assign its return value, `undefined`, to the `window.onload` property. Your `main` function starts executing immediately and tries to get the `root` div using `getElementById`. If your script tag is in the `head` element, the body of the document will not have been processed yet, therefore there is no such div to be found and `getElementById` will return `null`. Hence your code starts to fail. When the window 'load' event fires, nothing happens because the event handler is `undefined`.

If you place the script tag at the end of the body, then the root div will already have been created and everything seems to work fine, despite the `window.onload` event handler being `undefined`.

This is the correct code:

```js
 window.onload = () => main(HYF_REPOS_URL);
```

You have now _not_ called `main()` directly, but instead assigned an arrow function to the `window.onload` property that is to be called when the document is fully loaded (including the `body` with your root div). Now, it no longer matters where you place the script tag.